ICE: `bound types encountered in super_relate_tys`

<!--
Thank you for finding an Internal Compiler Error! 🧊  If possible, try to provide
a minimal verifiable example. You can read "Rust Bug Minimization Patterns" for
how to create smaller examples.
http://blog.pnkfx.org/blog/2019/11/18/rust-bug-minimization-patterns/
-->

### Code

```Rust
#![feature(non_lifetime_binders)]
#![crate_type = "lib"]

pub trait Lifetime<'a> {}

struct Foo2<Drop>
where
    for<'a> L: Lifetime<'a>,
{
    l: L,
}

impl<T: > Drop for Foo2<T>
where
    for<T: > T: Lifetime<'x>,
{
    fn drop(&mut self) {}
}
```


### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.71.0-nightly (c609da59d 2023-04-18)
binary: rustc
commit-hash: c609da59d9fc05b1c7dc879d79700ccd8140b5fc
commit-date: 2023-04-18
host: x86_64-unknown-linux-gnu
release: 1.71.0-nightly
LLVM version: 16.0.2
```

### Error output

```
error[E0261]: use of undeclared lifetime name `'x`
  --> treereduce.out:15:26
   |
15 |     for<T: > T: Lifetime<'x>,
   |                          ^^ undeclared lifetime
   |
   = note: for more information on higher-ranked polymorphism, visit https://doc.rust-lang.org/nomicon/hrtb.html
help: consider making the bound lifetime-generic with a new `'x` lifetime
   |
15 |     for<T: > T: for<'x> Lifetime<'x>,
   |                 +++++++
help: consider making the bound lifetime-generic with a new `'x` lifetime
   |
15 |     for<'x, T: > T: Lifetime<'x>,
   |         +++
help: consider introducing lifetime `'x` here
   |
13 | impl<'x, T: > Drop for Foo2<T>
   |      +++

error[E0412]: cannot find type `L` in this scope
 --> treereduce.out:8:13
  |
8 |     for<'a> L: Lifetime<'a>,
  |             ^ not found in this scope
  |
help: you might be missing a type parameter
  |
6 | struct Foo2<Drop, L>
  |                 +++

error[E0412]: cannot find type `L` in this scope
  --> treereduce.out:10:8
   |
10 |     l: L,
   |        ^ not found in this scope
   |
help: you might be missing a type parameter
   |
6  | struct Foo2<Drop, L>
   |                 +++

warning: the feature `non_lifetime_binders` is incomplete and may not be safe to use and/or cause compiler crashes
 --> treereduce.out:1:12
  |
1 | #![feature(non_lifetime_binders)]
  |            ^^^^^^^^^^^^^^^^^^^^
  |
  = note: see issue #108185 <https://github.com/rust-lang/rust/issues/108185> for more information
  = note: `#[warn(incomplete_features)]` on by default
```

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Backtrace</strong></summary>
<p>

```
error: internal compiler error: /rustc/c609da59d9fc05b1c7dc879d79700ccd8140b5fc/compiler/rustc_middle/src/ty/relate.rs:408:13: bound types encountered in super_relate_tys

thread 'rustc' panicked at 'Box<dyn Any>', /rustc/c609da59d9fc05b1c7dc879d79700ccd8140b5fc/compiler/rustc_errors/src/lib.rs:1643:9
stack backtrace:
   0:     0x7ff3e75653c2 - std::backtrace_rs::backtrace::libunwind::trace::h4acdb82da48c9833
                               at /rustc/c609da59d9fc05b1c7dc879d79700ccd8140b5fc/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x7ff3e75653c2 - std::backtrace_rs::backtrace::trace_unsynchronized::hf2cc6a3a410a5594
                               at /rustc/c609da59d9fc05b1c7dc879d79700ccd8140b5fc/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7ff3e75653c2 - std::sys_common::backtrace::_print_fmt::hbbc6914b749d0ada
                               at /rustc/c609da59d9fc05b1c7dc879d79700ccd8140b5fc/library/std/src/sys_common/backtrace.rs:65:5
   3:     0x7ff3e75653c2 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h3d5a35293018c1e6
                               at /rustc/c609da59d9fc05b1c7dc879d79700ccd8140b5fc/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x7ff3e75c95df - core::fmt::write::hcd8095c096d51cb1
                               at /rustc/c609da59d9fc05b1c7dc879d79700ccd8140b5fc/library/core/src/fmt/mod.rs:1254:17
   5:     0x7ff3e7558401 - std::io::Write::write_fmt::hfb074efb44accf99
                               at /rustc/c609da59d9fc05b1c7dc879d79700ccd8140b5fc/library/std/src/io/mod.rs:1698:15
   6:     0x7ff3e75651c5 - std::sys_common::backtrace::_print::hf3de088e34eea71b
                               at /rustc/c609da59d9fc05b1c7dc879d79700ccd8140b5fc/library/std/src/sys_common/backtrace.rs:47:5
   7:     0x7ff3e75651c5 - std::sys_common::backtrace::print::h8c1956d8c9acf6e3
                               at /rustc/c609da59d9fc05b1c7dc879d79700ccd8140b5fc/library/std/src/sys_common/backtrace.rs:34:9
   8:     0x7ff3e7567e87 - std::panicking::default_hook::{{closure}}::h0f8c867eadb4985a
   9:     0x7ff3e7567c75 - std::panicking::default_hook::h749ec9b15ecb3c39
                               at /rustc/c609da59d9fc05b1c7dc879d79700ccd8140b5fc/library/std/src/panicking.rs:288:9
  10:     0x7ff3ea830cf5 - rustc_driver_impl[7fdb2e8792375136]::DEFAULT_HOOK::{closure#0}::{closure#0}
  11:     0x7ff3e75685c5 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::ha6bcd52ce2118341
                               at /rustc/c609da59d9fc05b1c7dc879d79700ccd8140b5fc/library/alloc/src/boxed.rs:1987:9
  12:     0x7ff3e75685c5 - std::panicking::rust_panic_with_hook::h553363ec26086b73
                               at /rustc/c609da59d9fc05b1c7dc879d79700ccd8140b5fc/library/std/src/panicking.rs:695:13
  13:     0x7ff3ead18461 - std[aef50eda90d64d90]::panicking::begin_panic::<rustc_errors[934a790ce3691480]::ExplicitBug>::{closure#0}
  14:     0x7ff3ead17456 - std[aef50eda90d64d90]::sys_common::backtrace::__rust_end_short_backtrace::<std[aef50eda90d64d90]::panicking::begin_panic<rustc_errors[934a790ce3691480]::ExplicitBug>::{closure#0}, !>
  15:     0x7ff3ead5a9b6 - std[aef50eda90d64d90]::panicking::begin_panic::<rustc_errors[934a790ce3691480]::ExplicitBug>
  16:     0x7ff3ead83c26 - <rustc_errors[934a790ce3691480]::HandlerInner>::bug::<&alloc[3ca5cc4d7c388e]::string::String>
  17:     0x7ff3ead838f0 - <rustc_errors[934a790ce3691480]::Handler>::bug::<&alloc[3ca5cc4d7c388e]::string::String>
  18:     0x7ff3ead4463b - rustc_middle[8054a8be7b277c97]::util::bug::opt_span_bug_fmt::<rustc_span[8c04062e23e0b099]::span_encoding::Span>::{closure#0}
  19:     0x7ff3ead43f1a - rustc_middle[8054a8be7b277c97]::ty::context::tls::with_opt::<rustc_middle[8054a8be7b277c97]::util::bug::opt_span_bug_fmt<rustc_span[8c04062e23e0b099]::span_encoding::Span>::{closure#0}, !>::{closure#0}
  20:     0x7ff3ead43ee6 - rustc_middle[8054a8be7b277c97]::ty::context::tls::with_context_opt::<rustc_middle[8054a8be7b277c97]::ty::context::tls::with_opt<rustc_middle[8054a8be7b277c97]::util::bug::opt_span_bug_fmt<rustc_span[8c04062e23e0b099]::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
  21:     0x7ff3e89e6a7d - rustc_middle[8054a8be7b277c97]::util::bug::bug_fmt
  22:     0x7ff3ea03040a - rustc_middle[8054a8be7b277c97]::ty::relate::super_relate_tys::<rustc_hir_analysis[42144caad347631]::check::dropck::SimpleEqRelation>
  23:     0x7ff3e9e2f48f - <rustc_middle[8054a8be7b277c97]::ty::subst::GenericArg as rustc_middle[8054a8be7b277c97]::ty::relate::Relate>::relate::<rustc_hir_analysis[42144caad347631]::check::dropck::SimpleEqRelation>
  24:     0x7ff3e9e2f0b0 - <core[7c99772fdcabedb3]::result::Result<rustc_middle[8054a8be7b277c97]::ty::subst::GenericArg, rustc_middle[8054a8be7b277c97]::ty::error::TypeError> as rustc_type_ir[e1ddc2cff5d0a20e]::CollectAndApply<rustc_middle[8054a8be7b277c97]::ty::subst::GenericArg, &rustc_middle[8054a8be7b277c97]::ty::list::List<rustc_middle[8054a8be7b277c97]::ty::subst::GenericArg>>>::collect_and_apply::<core[7c99772fdcabedb3]::iter::adapters::map::Map<core[7c99772fdcabedb3]::iter::adapters::zip::Zip<core[7c99772fdcabedb3]::iter::adapters::copied::Copied<core[7c99772fdcabedb3]::slice::iter::Iter<rustc_middle[8054a8be7b277c97]::ty::subst::GenericArg>>, core[7c99772fdcabedb3]::iter::adapters::copied::Copied<core[7c99772fdcabedb3]::slice::iter::Iter<rustc_middle[8054a8be7b277c97]::ty::subst::GenericArg>>>, rustc_middle[8054a8be7b277c97]::ty::relate::relate_substs<rustc_hir_analysis[42144caad347631]::check::dropck::SimpleEqRelation>::{closure#0}>, <rustc_middle[8054a8be7b277c97]::ty::context::TyCtxt>::mk_substs_from_iter<core[7c99772fdcabedb3]::iter::adapters::map::Map<core[7c99772fdcabedb3]::iter::adapters::zip::Zip<core[7c99772fdcabedb3]::iter::adapters::copied::Copied<core[7c99772fdcabedb3]::slice::iter::Iter<rustc_middle[8054a8be7b277c97]::ty::subst::GenericArg>>, core[7c99772fdcabedb3]::iter::adapters::copied::Copied<core[7c99772fdcabedb3]::slice::iter::Iter<rustc_middle[8054a8be7b277c97]::ty::subst::GenericArg>>>, rustc_middle[8054a8be7b277c97]::ty::relate::relate_substs<rustc_hir_analysis[42144caad347631]::check::dropck::SimpleEqRelation>::{closure#0}>, core[7c99772fdcabedb3]::result::Result<rustc_middle[8054a8be7b277c97]::ty::subst::GenericArg, rustc_middle[8054a8be7b277c97]::ty::error::TypeError>>::{closure#0}>
  25:     0x7ff3e9e2ebd8 - <rustc_middle[8054a8be7b277c97]::ty::sty::TraitRef as rustc_middle[8054a8be7b277c97]::ty::relate::Relate>::relate::<rustc_hir_analysis[42144caad347631]::check::dropck::SimpleEqRelation>
  26:     0x7ff3e9e2e8a7 - <rustc_middle[8054a8be7b277c97]::ty::TraitPredicate as rustc_middle[8054a8be7b277c97]::ty::relate::Relate>::relate::<rustc_hir_analysis[42144caad347631]::check::dropck::SimpleEqRelation>
  27:     0x7ff3e9e2e57d - <core[7c99772fdcabedb3]::iter::adapters::copied::Copied<core[7c99772fdcabedb3]::slice::iter::Iter<rustc_middle[8054a8be7b277c97]::ty::Predicate>> as core[7c99772fdcabedb3]::iter::traits::iterator::Iterator>::try_fold::<(), core[7c99772fdcabedb3]::iter::traits::iterator::Iterator::any::check<rustc_middle[8054a8be7b277c97]::ty::Predicate, rustc_hir_analysis[42144caad347631]::check::dropck::ensure_drop_predicates_are_implied_by_item_defn::{closure#0}>::{closure#0}, core[7c99772fdcabedb3]::ops::control_flow::ControlFlow<()>>
  28:     0x7ff3e9e2d947 - rustc_hir_analysis[42144caad347631]::check::dropck::check_drop_impl
  29:     0x7ff3e9e2d2f7 - <rustc_middle[8054a8be7b277c97]::ty::context::TyCtxt>::calculate_dtor::<rustc_hir_analysis[42144caad347631]::check::dropck::check_drop_impl>::{closure#0}
  30:     0x7ff3e9e2ce21 - <rustc_middle[8054a8be7b277c97]::ty::context::TyCtxt>::calculate_dtor::<rustc_hir_analysis[42144caad347631]::check::dropck::check_drop_impl>
  31:     0x7ff3e9e2cb5b - rustc_hir_analysis[42144caad347631]::check::adt_destructor
  32:     0x7ff3e9cf5c1f - rustc_query_system[354c60129646c01e]::query::plumbing::try_execute_query::<rustc_query_impl[4ebfc2cbb59026c7]::queries::adt_destructor, rustc_query_impl[4ebfc2cbb59026c7]::plumbing::QueryCtxt>
  33:     0x7ff3e9cf5618 - <rustc_query_impl[4ebfc2cbb59026c7]::Queries as rustc_middle[8054a8be7b277c97]::ty::query::QueryEngine>::adt_destructor
  34:     0x7ff3e934594a - <rustc_middle[8054a8be7b277c97]::ty::adt::AdtDef>::destructor
  35:     0x7ff3e933e8f2 - rustc_hir_analysis[42144caad347631]::check::check::check_mod_item_types
  36:     0x7ff3e9e6fea1 - rustc_query_system[354c60129646c01e]::query::plumbing::try_execute_query::<rustc_query_impl[4ebfc2cbb59026c7]::queries::check_mod_item_types, rustc_query_impl[4ebfc2cbb59026c7]::plumbing::QueryCtxt>
  37:     0x7ff3e9e6fac9 - <rustc_query_impl[4ebfc2cbb59026c7]::Queries as rustc_middle[8054a8be7b277c97]::ty::query::QueryEngine>::check_mod_item_types
  38:     0x7ff3e9c0fbac - <rustc_middle[8054a8be7b277c97]::hir::map::Map>::for_each_module::<rustc_hir_analysis[42144caad347631]::check_crate::{closure#6}::{closure#0}>
  39:     0x7ff3e9c0bfab - rustc_hir_analysis[42144caad347631]::check_crate
  40:     0x7ff3e9c05e3d - rustc_interface[7925d28b3b6d3805]::passes::analysis
  41:     0x7ff3e9fa3712 - rustc_query_system[354c60129646c01e]::query::plumbing::try_execute_query::<rustc_query_impl[4ebfc2cbb59026c7]::queries::analysis, rustc_query_impl[4ebfc2cbb59026c7]::plumbing::QueryCtxt>
  42:     0x7ff3e9fa3420 - <rustc_query_impl[4ebfc2cbb59026c7]::Queries as rustc_middle[8054a8be7b277c97]::ty::query::QueryEngine>::analysis
  43:     0x7ff3e9e64ca6 - <std[aef50eda90d64d90]::thread::local::LocalKey<core[7c99772fdcabedb3]::cell::Cell<*const ()>>>::with::<rustc_middle[8054a8be7b277c97]::ty::context::tls::enter_context<<rustc_middle[8054a8be7b277c97]::ty::context::GlobalCtxt>::enter<rustc_driver_impl[7fdb2e8792375136]::run_compiler::{closure#1}::{closure#2}::{closure#4}, core[7c99772fdcabedb3]::result::Result<(), rustc_span[8c04062e23e0b099]::ErrorGuaranteed>>::{closure#0}, core[7c99772fdcabedb3]::result::Result<(), rustc_span[8c04062e23e0b099]::ErrorGuaranteed>>::{closure#0}, core[7c99772fdcabedb3]::result::Result<(), rustc_span[8c04062e23e0b099]::ErrorGuaranteed>>
  44:     0x7ff3e9a529a7 - <rustc_interface[7925d28b3b6d3805]::interface::Compiler>::enter::<rustc_driver_impl[7fdb2e8792375136]::run_compiler::{closure#1}::{closure#2}, core[7c99772fdcabedb3]::result::Result<core[7c99772fdcabedb3]::option::Option<rustc_interface[7925d28b3b6d3805]::queries::Linker>, rustc_span[8c04062e23e0b099]::ErrorGuaranteed>>
  45:     0x7ff3e9a4ca8f - rustc_span[8c04062e23e0b099]::set_source_map::<core[7c99772fdcabedb3]::result::Result<(), rustc_span[8c04062e23e0b099]::ErrorGuaranteed>, rustc_interface[7925d28b3b6d3805]::interface::run_compiler<core[7c99772fdcabedb3]::result::Result<(), rustc_span[8c04062e23e0b099]::ErrorGuaranteed>, rustc_driver_impl[7fdb2e8792375136]::run_compiler::{closure#1}>::{closure#0}::{closure#0}>
  46:     0x7ff3e9a4c6a8 - <scoped_tls[6d453f3239404e]::ScopedKey<rustc_span[8c04062e23e0b099]::SessionGlobals>>::set::<rustc_interface[7925d28b3b6d3805]::interface::run_compiler<core[7c99772fdcabedb3]::result::Result<(), rustc_span[8c04062e23e0b099]::ErrorGuaranteed>, rustc_driver_impl[7fdb2e8792375136]::run_compiler::{closure#1}>::{closure#0}, core[7c99772fdcabedb3]::result::Result<(), rustc_span[8c04062e23e0b099]::ErrorGuaranteed>>
  47:     0x7ff3e9a4bcbb - std[aef50eda90d64d90]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[7925d28b3b6d3805]::util::run_in_thread_pool_with_globals<rustc_interface[7925d28b3b6d3805]::interface::run_compiler<core[7c99772fdcabedb3]::result::Result<(), rustc_span[8c04062e23e0b099]::ErrorGuaranteed>, rustc_driver_impl[7fdb2e8792375136]::run_compiler::{closure#1}>::{closure#0}, core[7c99772fdcabedb3]::result::Result<(), rustc_span[8c04062e23e0b099]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[7c99772fdcabedb3]::result::Result<(), rustc_span[8c04062e23e0b099]::ErrorGuaranteed>>
  48:     0x7ff3ea101861 - <<std[aef50eda90d64d90]::thread::Builder>::spawn_unchecked_<rustc_interface[7925d28b3b6d3805]::util::run_in_thread_pool_with_globals<rustc_interface[7925d28b3b6d3805]::interface::run_compiler<core[7c99772fdcabedb3]::result::Result<(), rustc_span[8c04062e23e0b099]::ErrorGuaranteed>, rustc_driver_impl[7fdb2e8792375136]::run_compiler::{closure#1}>::{closure#0}, core[7c99772fdcabedb3]::result::Result<(), rustc_span[8c04062e23e0b099]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[7c99772fdcabedb3]::result::Result<(), rustc_span[8c04062e23e0b099]::ErrorGuaranteed>>::{closure#1} as core[7c99772fdcabedb3]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  49:     0x7ff3e7572a85 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::he0dee4d73e6ebe6e
                               at /rustc/c609da59d9fc05b1c7dc879d79700ccd8140b5fc/library/alloc/src/boxed.rs:1973:9
  50:     0x7ff3e7572a85 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::hdfd4a296008e148d
                               at /rustc/c609da59d9fc05b1c7dc879d79700ccd8140b5fc/library/alloc/src/boxed.rs:1973:9
  51:     0x7ff3e7572a85 - std::sys::unix::thread::Thread::new::thread_start::h8f06ac6e30e47e34
                               at /rustc/c609da59d9fc05b1c7dc879d79700ccd8140b5fc/library/std/src/sys/unix/thread.rs:108:17
  52:     0x7ff3e730dbb5 - <unknown>
  53:     0x7ff3e738fd90 - <unknown>
  54:                0x0 - <unknown>

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.71.0-nightly (c609da59d 2023-04-18) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [adt_destructor] computing `Drop` impl for `Foo2`
#1 [check_mod_item_types] checking item types in top-level module
#2 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 4 previous errors; 1 warning emitted

Some errors have detailed explanations: E0261, E0412.
For more information about an error, try `rustc --explain E0261`.
```

</p>
</details>

--- a/compiler/rustc_hir_analysis/src/check/dropck.rs
+++ b/compiler/rustc_hir_analysis/src/check/dropck.rs
@@ -1,12 +1,14 @@
 // FIXME(@lcnr): Move this module out of `rustc_hir_analysis`.
 //
 // We don't do any drop checking during hir typeck.
+use rustc_data_structures::fx::FxHashSet;
 use rustc_errors::{struct_span_err, ErrorGuaranteed};
-use rustc_middle::ty::error::TypeError;
-use rustc_middle::ty::relate::{Relate, RelateResult, TypeRelation};
+use rustc_infer::infer::outlives::env::OutlivesEnvironment;
+use rustc_infer::infer::{RegionResolutionError, TyCtxtInferExt};
 use rustc_middle::ty::subst::SubstsRef;
 use rustc_middle::ty::util::IgnoreRegions;
-use rustc_middle::ty::{self, Predicate, Ty, TyCtxt};
+use rustc_middle::ty::{self, TyCtxt};
+use rustc_trait_selection::traits::{self, ObligationCtxt};
 
 use crate::errors;
 use crate::hir::def_id::{DefId, LocalDefId};
@@ -43,21 +45,20 @@ pub fn check_drop_impl(tcx: TyCtxt<'_>, drop_impl_did: DefId) -> Result<(), Erro
         }
     }
     let dtor_self_type = tcx.type_of(drop_impl_did).subst_identity();
-    let dtor_predicates = tcx.predicates_of(drop_impl_did);
     match dtor_self_type.kind() {
-        ty::Adt(adt_def, self_to_impl_substs) => {
+        ty::Adt(adt_def, adt_to_impl_substs) => {
             ensure_drop_params_and_item_params_correspond(
                 tcx,
                 drop_impl_did.expect_local(),
                 adt_def.did(),
-                self_to_impl_substs,
+                adt_to_impl_substs,
             )?;
 
             ensure_drop_predicates_are_implied_by_item_defn(
                 tcx,
-                dtor_predicates,
+                drop_impl_did.expect_local(),
                 adt_def.did().expect_local(),
-                self_to_impl_substs,
+                adt_to_impl_substs,
             )
         }
         _ => {
@@ -78,9 +79,9 @@ fn ensure_drop_params_and_item_params_correspond<'tcx>(
     tcx: TyCtxt<'tcx>,
     drop_impl_did: LocalDefId,
     self_type_did: DefId,
-    drop_impl_substs: SubstsRef<'tcx>,
+    adt_to_impl_substs: SubstsRef<'tcx>,
 ) -> Result<(), ErrorGuaranteed> {
-    let Err(arg) = tcx.uses_unique_generic_params(drop_impl_substs, IgnoreRegions::No) else {
+    let Err(arg) = tcx.uses_unique_generic_params(adt_to_impl_substs, IgnoreRegions::No) else {
         return Ok(())
     };
 
@@ -111,237 +112,94 @@ fn ensure_drop_params_and_item_params_correspond<'tcx>(
 /// implied by assuming the predicates attached to self_type_did.
 fn ensure_drop_predicates_are_implied_by_item_defn<'tcx>(
     tcx: TyCtxt<'tcx>,
-    dtor_predicates: ty::GenericPredicates<'tcx>,
-    self_type_did: LocalDefId,
-    self_to_impl_substs: SubstsRef<'tcx>,
+    drop_impl_def_id: LocalDefId,
+    adt_def_id: LocalDefId,
+    adt_to_impl_substs: SubstsRef<'tcx>,
 ) -> Result<(), ErrorGuaranteed> {
-    let mut result = Ok(());
+    let infcx = tcx.infer_ctxt().build();
+    let ocx = ObligationCtxt::new(&infcx);
 
-    // Here is an example, analogous to that from
-    // `compare_impl_method`.
+    // Take the param-env of the adt and substitute the substs that show up in
+    // the implementation's self type. This gives us the assumptions that the
+    // self ty of the implementation is allowed to know just from it being a
+    // well-formed adt, since that's all we're allowed to assume while proving
+    // the Drop implementation is not specialized.
     //
-    // Consider a struct type:
-    //
-    //     struct Type<'c, 'b:'c, 'a> {
-    //         x: &'a Contents            // (contents are irrelevant;
-    //         y: &'c Cell<&'b Contents>, //  only the bounds matter for our purposes.)
-    //     }
-    //
-    // and a Drop impl:
-    //
-    //     impl<'z, 'y:'z, 'x:'y> Drop for P<'z, 'y, 'x> {
-    //         fn drop(&mut self) { self.y.set(self.x); } // (only legal if 'x: 'y)
-    //     }
-    //
-    // We start out with self_to_impl_substs, that maps the generic
-    // parameters of Type to that of the Drop impl.
-    //
-    //     self_to_impl_substs = {'c => 'z, 'b => 'y, 'a => 'x}
-    //
-    // Applying this to the predicates (i.e., assumptions) provided by the item
-    // definition yields the instantiated assumptions:
-    //
-    //     ['y : 'z]
-    //
-    // We then check all of the predicates of the Drop impl:
-    //
-    //     ['y:'z, 'x:'y]
-    //
-    // and ensure each is in the list of instantiated
-    // assumptions. Here, `'y:'z` is present, but `'x:'y` is
-    // absent. So we report an error that the Drop impl injected a
-    // predicate that is not present on the struct definition.
+    // We don't need to normalize this param-env or anything, since we're only
+    // substituting it with free params, so no additional param-env normalization
+    // can occur on top of what has been done in the param_env query itself.
+    let param_env = ty::EarlyBinder(tcx.param_env(adt_def_id))
+        .subst(tcx, adt_to_impl_substs)
+        .with_constness(tcx.constness(drop_impl_def_id));
 
-    // We can assume the predicates attached to struct/enum definition
-    // hold.
-    let generic_assumptions = tcx.predicates_of(self_type_did);
+    for (pred, span) in tcx.predicates_of(drop_impl_def_id).instantiate_identity(tcx) {
+        let normalize_cause = traits::ObligationCause::misc(span, adt_def_id);
+        let pred = ocx.normalize(&normalize_cause, param_env, pred);
+        let cause = traits::ObligationCause::new(span, adt_def_id, traits::DropImpl);
+        ocx.register_obligation(traits::Obligation::new(tcx, cause, param_env, pred));
+    }
 
-    let assumptions_in_impl_context = generic_assumptions.instantiate(tcx, &self_to_impl_substs);
-    let assumptions_in_impl_context = assumptions_in_impl_context.predicates;
+    // All of the custom error reporting logic is to preserve parity with the old
+    // error messages.
+    //
+    // They can probably get removed with better treatment of the new `DropImpl`
+    // obligation cause code, and perhaps some custom logic in `report_region_errors`.
 
-    debug!(?assumptions_in_impl_context, ?dtor_predicates.predicates);
-
-    let self_param_env = tcx.param_env(self_type_did);
-
-    // An earlier version of this code attempted to do this checking
-    // via the traits::fulfill machinery. However, it ran into trouble
-    // since the fulfill machinery merely turns outlives-predicates
-    // 'a:'b and T:'b into region inference constraints. It is simpler
-    // just to look for all the predicates directly.
-
-    assert_eq!(dtor_predicates.parent, None);
-    for &(predicate, predicate_sp) in dtor_predicates.predicates {
-        // (We do not need to worry about deep analysis of type
-        // expressions etc because the Drop impls are already forced
-        // to take on a structure that is roughly an alpha-renaming of
-        // the generic parameters of the item definition.)
-
-        // This path now just checks *all* predicates via an instantiation of
-        // the `SimpleEqRelation`, which simply forwards to the `relate` machinery
-        // after taking care of anonymizing late bound regions.
-        //
-        // However, it may be more efficient in the future to batch
-        // the analysis together via the fulfill (see comment above regarding
-        // the usage of the fulfill machinery), rather than the
-        // repeated `.iter().any(..)` calls.
-
-        // This closure is a more robust way to check `Predicate` equality
-        // than simple `==` checks (which were the previous implementation).
-        // It relies on `ty::relate` for `TraitPredicate`, `ProjectionPredicate`,
-        // `ConstEvaluatable` and `TypeOutlives` (which implement the Relate trait),
-        // while delegating on simple equality for the other `Predicate`.
-        // This implementation solves (Issue #59497) and (Issue #58311).
-        // It is unclear to me at the moment whether the approach based on `relate`
-        // could be extended easily also to the other `Predicate`.
-        let predicate_matches_closure = |p: Predicate<'tcx>| {
-            let mut relator: SimpleEqRelation<'tcx> = SimpleEqRelation::new(tcx, self_param_env);
-            let predicate = predicate.kind();
-            let p = p.kind();
-            match (predicate.skip_binder(), p.skip_binder()) {
-                (
-                    ty::PredicateKind::Clause(ty::Clause::Trait(a)),
-                    ty::PredicateKind::Clause(ty::Clause::Trait(b)),
-                ) => relator.relate(predicate.rebind(a), p.rebind(b)).is_ok(),
-                (
-                    ty::PredicateKind::Clause(ty::Clause::Projection(a)),
-                    ty::PredicateKind::Clause(ty::Clause::Projection(b)),
-                ) => relator.relate(predicate.rebind(a), p.rebind(b)).is_ok(),
-                (
-                    ty::PredicateKind::ConstEvaluatable(a),
-                    ty::PredicateKind::ConstEvaluatable(b),
-                ) => relator.relate(predicate.rebind(a), predicate.rebind(b)).is_ok(),
-                (
-                    ty::PredicateKind::Clause(ty::Clause::TypeOutlives(ty::OutlivesPredicate(
-                        ty_a,
-                        lt_a,
-                    ))),
-                    ty::PredicateKind::Clause(ty::Clause::TypeOutlives(ty::OutlivesPredicate(
-                        ty_b,
-                        lt_b,
-                    ))),
-                ) => {
-                    relator.relate(predicate.rebind(ty_a), p.rebind(ty_b)).is_ok()
-                        && relator.relate(predicate.rebind(lt_a), p.rebind(lt_b)).is_ok()
-                }
-                (ty::PredicateKind::WellFormed(arg_a), ty::PredicateKind::WellFormed(arg_b)) => {
-                    relator.relate(predicate.rebind(arg_a), p.rebind(arg_b)).is_ok()
-                }
-                _ => predicate == p,
+    let errors = ocx.select_all_or_error();
+    if !errors.is_empty() {
+        let mut guar = None;
+        let mut root_predicates = FxHashSet::default();
+        for error in errors {
+            let root_predicate = error.root_obligation.predicate;
+            if root_predicates.insert(root_predicate) {
+                let item_span = tcx.def_span(adt_def_id);
+                let self_descr = tcx.def_descr(adt_def_id.to_def_id());
+                guar = Some(
+                    struct_span_err!(
+                        tcx.sess,
+                        error.root_obligation.cause.span,
+                        E0367,
+                        "`Drop` impl requires `{root_predicate}` \
+                        but the {self_descr} it is implemented for does not",
+                    )
+                    .span_note(item_span, "the implementor must specify the same requirement")
+                    .emit(),
+                );
             }
-        };
-
-        if !assumptions_in_impl_context.iter().copied().any(predicate_matches_closure) {
-            let item_span = tcx.def_span(self_type_did);
-            let self_descr = tcx.def_descr(self_type_did.to_def_id());
-            let reported = struct_span_err!(
-                tcx.sess,
-                predicate_sp,
-                E0367,
-                "`Drop` impl requires `{predicate}` but the {self_descr} it is implemented for does not",
-            )
-            .span_note(item_span, "the implementor must specify the same requirement")
-            .emit();
-            result = Err(reported);
         }
+        return Err(guar.unwrap());
     }
 
-    result
-}
-
-/// This is an implementation of the [`TypeRelation`] trait with the
-/// aim of simply comparing for equality (without side-effects).
-///
-/// It is not intended to be used anywhere else other than here.
-pub(crate) struct SimpleEqRelation<'tcx> {
-    tcx: TyCtxt<'tcx>,
-    param_env: ty::ParamEnv<'tcx>,
-}
-
-impl<'tcx> SimpleEqRelation<'tcx> {
-    fn new(tcx: TyCtxt<'tcx>, param_env: ty::ParamEnv<'tcx>) -> SimpleEqRelation<'tcx> {
-        SimpleEqRelation { tcx, param_env }
-    }
-}
-
-impl<'tcx> TypeRelation<'tcx> for SimpleEqRelation<'tcx> {
-    fn tcx(&self) -> TyCtxt<'tcx> {
-        self.tcx
-    }
-
-    fn param_env(&self) -> ty::ParamEnv<'tcx> {
-        self.param_env
-    }
-
-    fn tag(&self) -> &'static str {
-        "dropck::SimpleEqRelation"
-    }
-
-    fn a_is_expected(&self) -> bool {
-        true
-    }
-
-    fn relate_with_variance<T: Relate<'tcx>>(
-        &mut self,
-        _: ty::Variance,
-        _info: ty::VarianceDiagInfo<'tcx>,
-        a: T,
-        b: T,
-    ) -> RelateResult<'tcx, T> {
-        // Here we ignore variance because we require drop impl's types
-        // to be *exactly* the same as to the ones in the struct definition.
-        self.relate(a, b)
-    }
-
-    fn tys(&mut self, a: Ty<'tcx>, b: Ty<'tcx>) -> RelateResult<'tcx, Ty<'tcx>> {
-        debug!("SimpleEqRelation::tys(a={:?}, b={:?})", a, b);
-        ty::relate::super_relate_tys(self, a, b)
-    }
-
-    fn regions(
-        &mut self,
-        a: ty::Region<'tcx>,
-        b: ty::Region<'tcx>,
-    ) -> RelateResult<'tcx, ty::Region<'tcx>> {
-        debug!("SimpleEqRelation::regions(a={:?}, b={:?})", a, b);
-
-        // We can just equate the regions because LBRs have been
-        // already anonymized.
-        if a == b {
-            Ok(a)
-        } else {
-            // I'm not sure is this `TypeError` is the right one, but
-            // it should not matter as it won't be checked (the dropck
-            // will emit its own, more informative and higher-level errors
-            // in case anything goes wrong).
-            Err(TypeError::RegionsPlaceholderMismatch)
+    let errors = ocx.infcx.resolve_regions(&OutlivesEnvironment::new(param_env));
+    if !errors.is_empty() {
+        let mut guar = None;
+        for error in errors {
+            let item_span = tcx.def_span(adt_def_id);
+            let self_descr = tcx.def_descr(adt_def_id.to_def_id());
+            let outlives = match error {
+                RegionResolutionError::ConcreteFailure(_, a, b) => format!("{b}: {a}"),
+                RegionResolutionError::GenericBoundFailure(_, generic, r) => {
+                    format!("{generic}: {r}")
+                }
+                RegionResolutionError::SubSupConflict(_, _, _, a, _, b, _) => format!("{b}: {a}"),
+                RegionResolutionError::UpperBoundUniverseConflict(a, _, _, _, b) => {
+                    format!("{b}: {a}", a = tcx.mk_re_var(a))
+                }
+            };
+            guar = Some(
+                struct_span_err!(
+                    tcx.sess,
+                    error.origin().span(),
+                    E0367,
+                    "`Drop` impl requires `{outlives}` \
+                    but the {self_descr} it is implemented for does not",
+                )
+                .span_note(item_span, "the implementor must specify the same requirement")
+                .emit(),
+            );
         }
+        return Err(guar.unwrap());
     }
 
-    fn consts(
-        &mut self,
-        a: ty::Const<'tcx>,
-        b: ty::Const<'tcx>,
-    ) -> RelateResult<'tcx, ty::Const<'tcx>> {
-        debug!("SimpleEqRelation::consts(a={:?}, b={:?})", a, b);
-        ty::relate::super_relate_consts(self, a, b)
-    }
-
-    fn binders<T>(
-        &mut self,
-        a: ty::Binder<'tcx, T>,
-        b: ty::Binder<'tcx, T>,
-    ) -> RelateResult<'tcx, ty::Binder<'tcx, T>>
-    where
-        T: Relate<'tcx>,
-    {
-        debug!("SimpleEqRelation::binders({:?}: {:?}", a, b);
-
-        // Anonymizing the LBRs is necessary to solve (Issue #59497).
-        // After we do so, it should be totally fine to skip the binders.
-        let anon_a = self.tcx.anonymize_bound_vars(a);
-        let anon_b = self.tcx.anonymize_bound_vars(b);
-        self.relate(anon_a.skip_binder(), anon_b.skip_binder())?;
-
-        Ok(a)
-    }
+    Ok(())
 }

--- a/compiler/rustc_infer/src/infer/lexical_region_resolve/mod.rs
+++ b/compiler/rustc_infer/src/infer/lexical_region_resolve/mod.rs
@@ -102,6 +102,17 @@ pub enum RegionResolutionError<'tcx> {
     ),
 }
 
+impl<'tcx> RegionResolutionError<'tcx> {
+    pub fn origin(&self) -> &SubregionOrigin<'tcx> {
+        match self {
+            RegionResolutionError::ConcreteFailure(origin, _, _)
+            | RegionResolutionError::GenericBoundFailure(origin, _, _)
+            | RegionResolutionError::SubSupConflict(_, _, origin, _, _, _, _)
+            | RegionResolutionError::UpperBoundUniverseConflict(_, _, _, origin, _) => origin,
+        }
+    }
+}
+
 struct RegionAndOrigin<'tcx> {
     region: Region<'tcx>,
     origin: SubregionOrigin<'tcx>,

--- a/compiler/rustc_middle/src/traits/mod.rs
+++ b/compiler/rustc_middle/src/traits/mod.rs
@@ -444,6 +444,10 @@ pub enum ObligationCauseCode<'tcx> {
     AscribeUserTypeProvePredicate(Span),
 
     RustCall,
+
+    /// Obligations to prove that a `std::ops::Drop` impl is not stronger than
+    /// the ADT it's being implemented for.
+    DropImpl,
 }
 
 /// The 'location' at which we try to perform HIR-based wf checking.

--- a/compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs
+++ b/compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs
@@ -2790,7 +2790,8 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
             | ObligationCauseCode::LetElse
             | ObligationCauseCode::BinOp { .. }
             | ObligationCauseCode::AscribeUserTypeProvePredicate(..)
-            | ObligationCauseCode::RustCall => {}
+            | ObligationCauseCode::RustCall
+            | ObligationCauseCode::DropImpl => {}
             ObligationCauseCode::SliceOrArrayElem => {
                 err.note("slice and array elements must have `Sized` type");
             }

--- a/tests/ui/dropck/explicit-drop-bounds.bad1.stderr
+++ b/tests/ui/dropck/explicit-drop-bounds.bad1.stderr
@@ -1,0 +1,35 @@
+error[E0277]: the trait bound `T: Copy` is not satisfied
+  --> $DIR/explicit-drop-bounds.rs:27:18
+   |
+LL | impl<T> Drop for DropMe<T>
+   |                  ^^^^^^^^^ the trait `Copy` is not implemented for `T`
+   |
+note: required by a bound in `DropMe`
+  --> $DIR/explicit-drop-bounds.rs:7:18
+   |
+LL | struct DropMe<T: Copy>(T);
+   |                  ^^^^ required by this bound in `DropMe`
+help: consider further restricting type parameter `T`
+   |
+LL |     [T; 1]: Copy, T: std::marker::Copy // But `[T; 1]: Copy` does not imply `T: Copy`
+   |                 ~~~~~~~~~~~~~~~~~~~~~~
+
+error[E0277]: the trait bound `T: Copy` is not satisfied
+  --> $DIR/explicit-drop-bounds.rs:32:13
+   |
+LL |     fn drop(&mut self) {}
+   |             ^^^^^^^^^ the trait `Copy` is not implemented for `T`
+   |
+note: required by a bound in `DropMe`
+  --> $DIR/explicit-drop-bounds.rs:7:18
+   |
+LL | struct DropMe<T: Copy>(T);
+   |                  ^^^^ required by this bound in `DropMe`
+help: consider further restricting type parameter `T`
+   |
+LL |     [T; 1]: Copy, T: std::marker::Copy // But `[T; 1]: Copy` does not imply `T: Copy`
+   |                 ~~~~~~~~~~~~~~~~~~~~~~
+
+error: aborting due to 2 previous errors
+
+For more information about this error, try `rustc --explain E0277`.

--- a/tests/ui/dropck/explicit-drop-bounds.bad2.stderr
+++ b/tests/ui/dropck/explicit-drop-bounds.bad2.stderr
@@ -1,0 +1,35 @@
+error[E0277]: the trait bound `T: Copy` is not satisfied
+  --> $DIR/explicit-drop-bounds.rs:37:18
+   |
+LL | impl<T> Drop for DropMe<T>
+   |                  ^^^^^^^^^ the trait `Copy` is not implemented for `T`
+   |
+note: required by a bound in `DropMe`
+  --> $DIR/explicit-drop-bounds.rs:7:18
+   |
+LL | struct DropMe<T: Copy>(T);
+   |                  ^^^^ required by this bound in `DropMe`
+help: consider restricting type parameter `T`
+   |
+LL | impl<T: std::marker::Copy> Drop for DropMe<T>
+   |       +++++++++++++++++++
+
+error[E0277]: the trait bound `T: Copy` is not satisfied
+  --> $DIR/explicit-drop-bounds.rs:40:13
+   |
+LL |     fn drop(&mut self) {}
+   |             ^^^^^^^^^ the trait `Copy` is not implemented for `T`
+   |
+note: required by a bound in `DropMe`
+  --> $DIR/explicit-drop-bounds.rs:7:18
+   |
+LL | struct DropMe<T: Copy>(T);
+   |                  ^^^^ required by this bound in `DropMe`
+help: consider restricting type parameter `T`
+   |
+LL | impl<T: std::marker::Copy> Drop for DropMe<T>
+   |       +++++++++++++++++++
+
+error: aborting due to 2 previous errors
+
+For more information about this error, try `rustc --explain E0277`.

--- a/tests/ui/dropck/explicit-drop-bounds.rs
+++ b/tests/ui/dropck/explicit-drop-bounds.rs
@@ -1,0 +1,44 @@
+// revisions: good1 good2 bad1 bad2
+//[good1] check-pass
+//[good2] check-pass
+
+use std::ops::Drop;
+
+struct DropMe<T: Copy>(T);
+
+#[cfg(good1)]
+impl<T> Drop for DropMe<T>
+where
+    T: Copy + Clone,
+{
+    fn drop(&mut self) {}
+}
+
+#[cfg(good2)]
+impl<T> Drop for DropMe<T>
+where
+    T: Copy,
+    [T; 1]: Copy, // Trivial bound implied by `T: Copy`
+{
+    fn drop(&mut self) {}
+}
+
+#[cfg(bad1)]
+impl<T> Drop for DropMe<T>
+//[bad1]~^ ERROR the trait bound `T: Copy` is not satisfied
+where
+    [T; 1]: Copy, // But `[T; 1]: Copy` does not imply `T: Copy`
+{
+    fn drop(&mut self) {}
+    //[bad1]~^ ERROR the trait bound `T: Copy` is not satisfied
+}
+
+#[cfg(bad2)]
+impl<T> Drop for DropMe<T>
+//[bad2]~^ ERROR the trait bound `T: Copy` is not satisfied
+{
+    fn drop(&mut self) {}
+    //[bad2]~^ ERROR the trait bound `T: Copy` is not satisfied
+}
+
+fn main() {}

--- a/tests/ui/dropck/explicit-implied-outlives.bad1.stderr
+++ b/tests/ui/dropck/explicit-implied-outlives.bad1.stderr
@@ -1,0 +1,15 @@
+error[E0367]: `Drop` impl requires `T: 'static` but the struct it is implemented for does not
+  --> $DIR/explicit-implied-outlives.rs:28:8
+   |
+LL |     T: 'static,
+   |        ^^^^^^^
+   |
+note: the implementor must specify the same requirement
+  --> $DIR/explicit-implied-outlives.rs:7:1
+   |
+LL | struct DropMe<'a, T>(&'a T);
+   | ^^^^^^^^^^^^^^^^^^^^
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0367`.

--- a/tests/ui/dropck/explicit-implied-outlives.bad2.stderr
+++ b/tests/ui/dropck/explicit-implied-outlives.bad2.stderr
@@ -1,0 +1,15 @@
+error[E0367]: `Drop` impl requires `'a: 'static` but the struct it is implemented for does not
+  --> $DIR/explicit-implied-outlives.rs:37:9
+   |
+LL |     'a: 'static,
+   |         ^^^^^^^
+   |
+note: the implementor must specify the same requirement
+  --> $DIR/explicit-implied-outlives.rs:7:1
+   |
+LL | struct DropMe<'a, T>(&'a T);
+   | ^^^^^^^^^^^^^^^^^^^^
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0367`.

--- a/tests/ui/dropck/explicit-implied-outlives.rs
+++ b/tests/ui/dropck/explicit-implied-outlives.rs
@@ -1,0 +1,43 @@
+// revisions: good1 good2 bad1 bad2
+//[good1] check-pass
+//[good2] check-pass
+
+use std::ops::Drop;
+
+struct DropMe<'a, T>(&'a T);
+
+#[cfg(good1)]
+impl<'a, T> Drop for DropMe<'a, T>
+where
+    T: 'a, // Implied by struct, explicit on impl
+{
+    fn drop(&mut self) {}
+}
+
+#[cfg(good2)]
+impl<'a, T> Drop for DropMe<'a, T>
+where
+    'static: 'a, // Trivial bound
+{
+    fn drop(&mut self) {}
+}
+
+#[cfg(bad1)]
+impl<'a, T> Drop for DropMe<'a, T>
+where
+    T: 'static,
+    //[bad1]~^ ERROR `Drop` impl requires `T: 'static`
+{
+    fn drop(&mut self) {}
+}
+
+#[cfg(bad2)]
+impl<'a, T> Drop for DropMe<'a, T>
+where
+    'a: 'static,
+    //[bad2]~^ ERROR `Drop` impl requires `'a: 'static`
+{
+    fn drop(&mut self) {}
+}
+
+fn main() {}

--- a/tests/ui/dropck/transitive-outlives-2.rs
+++ b/tests/ui/dropck/transitive-outlives-2.rs
@@ -1,0 +1,18 @@
+// check-pass
+
+use std::marker::PhantomData;
+use std::ops::Drop;
+
+// a >= b >= c >= a implies a = b = c
+struct DropMe<'a: 'b, 'b: 'c, 'c: 'a>(
+    PhantomData<&'a ()>,
+    PhantomData<&'b ()>,
+    PhantomData<&'c ()>,
+);
+
+// a >= b, a >= c, b >= a, c >= a implies a = b = c
+impl<'a: 'b + 'c, 'b: 'a, 'c: 'a> Drop for DropMe<'a, 'b, 'c> {
+    fn drop(&mut self) {}
+}
+
+fn main() {}

--- a/tests/ui/dropck/transitive-outlives.bad.stderr
+++ b/tests/ui/dropck/transitive-outlives.bad.stderr
@@ -1,0 +1,15 @@
+error[E0367]: `Drop` impl requires `'a: 'c` but the struct it is implemented for does not
+  --> $DIR/transitive-outlives.rs:20:9
+   |
+LL |     'a: 'c,
+   |         ^^
+   |
+note: the implementor must specify the same requirement
+  --> $DIR/transitive-outlives.rs:7:1
+   |
+LL | struct DropMe<'a, 'b: 'a, 'c: 'b>(PhantomData<&'a ()>, PhantomData<&'b ()>, PhantomData<&'c ()>);
+   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0367`.

--- a/tests/ui/dropck/transitive-outlives.rs
+++ b/tests/ui/dropck/transitive-outlives.rs
@@ -1,0 +1,26 @@
+// revisions: good bad
+//[good] check-pass
+
+use std::marker::PhantomData;
+use std::ops::Drop;
+
+struct DropMe<'a, 'b: 'a, 'c: 'b>(PhantomData<&'a ()>, PhantomData<&'b ()>, PhantomData<&'c ()>);
+
+#[cfg(good)]
+impl<'a, 'b, 'c> Drop for DropMe<'a, 'b, 'c>
+where
+    'c: 'a,
+{
+    fn drop(&mut self) {}
+}
+
+#[cfg(bad)]
+impl<'a, 'b, 'c> Drop for DropMe<'a, 'b, 'c>
+where
+    'a: 'c,
+    //[bad]~^ ERROR `Drop` impl requires `'a: 'c`
+{
+    fn drop(&mut self) {}
+}
+
+fn main() {}

--- a/tests/ui/dropck/trivial-impl-bounds.rs
+++ b/tests/ui/dropck/trivial-impl-bounds.rs
@@ -1,0 +1,34 @@
+// revisions: good1 good2 good3
+// check-pass
+
+use std::ops::Drop;
+
+struct Foo;
+
+const X: usize = 1;
+
+#[cfg(good1)]
+impl Drop for Foo
+where
+    [(); X]:, // Trivial WF bound
+{
+    fn drop(&mut self) {}
+}
+
+#[cfg(good2)]
+impl Drop for Foo
+where
+    for<'a> &'a (): Copy, // Trivial trait bound
+{
+    fn drop(&mut self) {}
+}
+
+#[cfg(good3)]
+impl Drop for Foo
+where
+    for<'a> &'a (): 'a, // Trivial outlives bound
+{
+    fn drop(&mut self) {}
+}
+
+fn main() {}

--- a/tests/ui/traits/non_lifetime_binders/drop-impl-pred.no.stderr
+++ b/tests/ui/traits/non_lifetime_binders/drop-impl-pred.no.stderr
@@ -1,0 +1,24 @@
+warning: the feature `non_lifetime_binders` is incomplete and may not be safe to use and/or cause compiler crashes
+  --> $DIR/drop-impl-pred.rs:6:12
+   |
+LL | #![feature(non_lifetime_binders)]
+   |            ^^^^^^^^^^^^^^^^^^^^
+   |
+   = note: see issue #108185 <https://github.com/rust-lang/rust/issues/108185> for more information
+   = note: `#[warn(incomplete_features)]` on by default
+
+error[E0367]: `Drop` impl requires `H: Foo` but the struct it is implemented for does not
+  --> $DIR/drop-impl-pred.rs:19:15
+   |
+LL |     for<H> H: Foo,
+   |               ^^^
+   |
+note: the implementor must specify the same requirement
+  --> $DIR/drop-impl-pred.rs:12:1
+   |
+LL | struct Bar<T>(T) where T: Foo;
+   | ^^^^^^^^^^^^^
+
+error: aborting due to previous error; 1 warning emitted
+
+For more information about this error, try `rustc --explain E0367`.

--- a/tests/ui/traits/non_lifetime_binders/drop-impl-pred.rs
+++ b/tests/ui/traits/non_lifetime_binders/drop-impl-pred.rs
@@ -1,0 +1,25 @@
+// revisions: no yes
+//[yes] check-pass
+
+// Issue 110557
+
+#![feature(non_lifetime_binders)]
+//~^ WARN the feature `non_lifetime_binders` is incomplete
+
+pub trait Foo {}
+
+#[cfg(no)]
+struct Bar<T>(T) where T: Foo;
+
+#[cfg(yes)]
+struct Bar<T>(T) where for<H> H: Foo;
+
+impl<T> Drop for Bar<T>
+where
+    for<H> H: Foo,
+//[no]~^ ERROR `Drop` impl requires `H: Foo` but the struct it is implemented for does not
+{
+    fn drop(&mut self) {}
+}
+
+fn main() {}

--- a/tests/ui/traits/non_lifetime_binders/drop-impl-pred.yes.stderr
+++ b/tests/ui/traits/non_lifetime_binders/drop-impl-pred.yes.stderr
@@ -1,0 +1,11 @@
+warning: the feature `non_lifetime_binders` is incomplete and may not be safe to use and/or cause compiler crashes
+  --> $DIR/drop-impl-pred.rs:6:12
+   |
+LL | #![feature(non_lifetime_binders)]
+   |            ^^^^^^^^^^^^^^^^^^^^
+   |
+   = note: see issue #108185 <https://github.com/rust-lang/rust/issues/108185> for more information
+   = note: `#[warn(incomplete_features)]` on by default
+
+warning: 1 warning emitted
+

